File: src/main/java/com/pricetra/email_server/controllers/PasswordResetController.java
package com.pricetra.email_server.controllers;

import com.pricetra.email_server.dto.PasswordResetRequest;
import com.pricetra.email_server.response.PasswordResetResponse;
import com.sendgrid.Response;
import com.sendgrid.helpers.mail.Mail;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RestController;

import java.io.IOException;

@RestController
public class PasswordResetController extends BaseController {
    @PostMapping("password-reset")
    public PasswordResetResponse sendPasswordResetCode(@RequestBody PasswordResetRequest params) throws IOException {
        String html = getTemplateAsString("password-reset.html")
                .replace("{{fullName}}", params.fullName)
                .replace("{{avatarUrl}}", params.avatarUrl)
                .replace("{{code}}", params.code);
        Mail mail = this.newHtmlMailer(noReplyEmail, params.recipientEmail, "Password Reset Code", html);
        Response emailResponse = this.sendEmail(mail);
        PasswordResetResponse res = new PasswordResetResponse();
        res.content = html;
        res.status = Integer.toString(emailResponse.getStatusCode());
        res.code = params.code;
        res.fullName = params.fullName;
        res.avatarUrl = params.avatarUrl;
        res.subject = mail.subject;
        res.recipientEmail = params.recipientEmail;
        return res;
    }
}

File: src/controllers/password-reset-controller.ts
import { Body, Post, Route, Tags, SuccessResponse } from "tsoa";
import type { PasswordResetRequest, PasswordResetResponse } from "../models/password-reset";
import { fillTemplateVariables, getEmailTemplate } from "../utils/mail-helpers";
import { BaseController } from "./controller";

@Route("password-reset")
  @Tags("password-reset-controller")
export class PasswordResetController extends BaseController {
  @Post()
  @SuccessResponse("200", "OK")
  public async sendPasswordResetCode(@Body() body: PasswordResetRequest): Promise<PasswordResetResponse> {
    const vars = new Map([
      ['code', body.code],
      ['fullName', body.fullName],
    ]);
    if (body.avatarUrl) vars.set('avatarUrl', body.avatarUrl);
    const html = fillTemplateVariables(getEmailTemplate('password-reset'), vars);
    const subject = 'Password Reset Code';
    const { error } = await this.sendNoReplyEmail(body.recipientEmail, subject, html);
    if (error) throw new Error(`Resend Error: ${error.name} - ${error.message}`);

    return {
      content: html,
      status: 200,
      subject,
      ...body,
    };
  }
}
